### ISSUE TITLE
Calling tf.keras.models.load_model / model.save twice fails

### ISSUE THREAD

**Issue by srfrnk** [tensorflow/tensorflow#33085]
Please see [CoLab](https://colab.research.google.com/drive/1dXgiJZDbGHdQFrquvtMBblgz_5O1WSlD) or [Github Gist](https://gist.github.com/srfrnk/713765b44dd6677ba3e26ff4ac46f68b#file-save-load-bug-ipynb) to see live demo.

Steps to reproduce:
1) Create model,compile and fit
2) Save model
3) Load model and fit
4) Save model
5) Load model

Get error: `ValueError: Passing a dictionary input to a Sequential Model which doesn't have FeatureLayer as the first layer is an error.`

Expected: Model that was loaded and saved again should be loaded without error.

**Comment by rmothukuru** [comment]
@srfrnk,
I have executed your code in Tensorflow Version 2.0 and it worked fine. Here is the [Gist](https://colab.sandbox.google.com/gist/rmothukuru/602f614b31dba5ef5e98026d317caa89/33085.ipynb).

Thanks!

**Comment by srfrnk** [comment]
I have tested again using this [CoLab](https://colab.research.google.com/drive/1dXgiJZDbGHdQFrquvtMBblgz_5O1WSlD)

The issue is reproduced. You can verify by running the CoLab yourself.
Since CoLab runs remotely I hardly think it can be something in my local environment.

I also printed the output of the CoLab here:
[Save-Load-Bug - Colaboratory.pdf](https://github.com/tensorflow/tensorflow/files/3696934/Save-Load-Bug.-.Colaboratory.pdf)

WDYT?

**Comment by rmothukuru** [comment]
@srfrnk,
Tried executing the Colab but facing the error, `OSError: Unable to create file (unable to open file: name = './data/titanic_model', errno = 2, error message = 'No such file or directory', flags = 13, o_flags = 242) `.

Can you please provide complete information so that we can reproduce it at our end. Thanks!

**Comment by srfrnk** [comment]
@rmothukuru 
Are you sure you executed the steps in order and didn't miss any?
The error you are facing indicates trying to read the files at step 5 with `model = tf.keras.models.load_model('./data/titanic_model')`
However the files should have been created by step 4 with `model.save('./data/titanic_model')`
Perhaps you can check that step 4 executed correctly?

**Comment by rmothukuru** [comment]
@srfrnk,
Thank you for your input. Now I could run your colab without any error. Please find the [Gist](https://colab.sandbox.google.com/gist/rmothukuru/bfaa5d4418f49947c41fc467fc50c3be/33085.ipynb). Can you please recheck at your end. Thanks!

**Comment by srfrnk** [comment]
I am getting the same results still - it fails with the CoLab I create.
However the CoLab you sent passes for me - yet I can't see any difference.
It's very strange as CoLab should behave the same for both of us as it is not run locally.
I am trying to find another (maybe more stable) environment that can replicate the issue.

**Comment by rmothukuru** [comment]
@srfrnk,
Can you please confirm if you have restarted your Run Time of Google Colab and if Possible, restart your machine and check if the issue still persists. Thanks!

**Comment by srfrnk** [comment]
I can confirm having done so several times.
I've tried something new.
First I've switched from `model.save` to the static `tf.keras.models.save_model` to be symmetric to the loading mechanism
Also I combined the cells into one and renamed the model filepath.
Maybe that would help to make a more stable test case.

Hopefully this time we can both get the same output for the same input.
If that won't help I can think of using a docker container... WDYT?

## Steps to reproduce:
Go to [CoLab](https://colab.research.google.com) and click the following button at the bottom right corner:
![screenshot-colab research google com-2019 10 11-10-38-10](https://user-images.githubusercontent.com/1501654/66633316-3b018580-ec13-11e9-9542-7dc453abf34a.png)

Within the new notebook that opens - paste the following code inside single cell:
```
!pip3 install --user tensorflow==2.0.0 numpy==1.17.2 gast==0.2.2 matplotlib pandas pathlib pyyaml h5py
!rm -rf ./model
from __future__ import absolute_import, division, print_function, unicode_literals

import os
import tensorflow as tf
import numpy as np
import pathlib
import matplotlib.pyplot as plt
import pandas as pd
from tensorflow import keras
from tensorflow.keras import layers
import functools
from collections import OrderedDict

print(tf.version.VERSION)
print(tf.keras.__version__)
print(np.__version__)

pd.set_option('display.max_rows', 1000)

ds = tf.data.Dataset.from_tensor_slices(
    (OrderedDict([('Feature',tf.ones(shape=(1000)))]),tf.ones(shape=(1000)))
).batch(10).repeat()

model = tf.keras.Sequential([
    layers.DenseFeatures([tf.feature_column.numeric_column('Feature')], name='features'),
    layers.Dense(100, activation='relu', name='hidden3'),
    layers.Dense(1, activation='sigmoid', name='outputs')
])

model.compile(optimizer=keras.optimizers.Adam(),
              loss=keras.losses.BinaryCrossentropy(),
              metrics=[keras.metrics.BinaryAccuracy()])
model.fit(ds, epochs=1, steps_per_epoch=10)
tf.keras.models.save_model(
    model=model,
    filepath='./model',
    overwrite=True,
    include_optimizer=True,
    save_format='tf')

model = tf.keras.models.load_model('./model')
model.fit(ds, epochs=1, steps_per_epoch=10)
tf.keras.models.save_model(
    model=model,
    filepath='./model',
    overwrite=True,
    include_optimizer=True,
    save_format='tf')

model = tf.keras.models.load_model('./model')
model.fit(ds, epochs=1, steps_per_epoch=10)
tf.keras.models.save_model(
    model=model,
    filepath='./model',
    overwrite=True,
    include_optimizer=True,
    save_format='tf')
```


Run the cell, wait until it shows the following and click the button :
![screenshot-colab research google com-2019 10 11-10-41-48](https://user-images.githubusercontent.com/1501654/66633569-c7ac4380-ec13-11e9-8fdc-6d1bb4045e80.png)

After notebook reconnects, run the cell again until completion.

You should see he following:
![screenshot-colab research google com-2019 10 11-10-56-50](https://user-images.githubusercontent.com/1501654/66634493-ddbb0380-ec15-11e9-89a5-9422a1f81824.png)

**Comment by rmothukuru** [comment]
@srfrnk,
I could reproduce the issue now. Can you please refer [this link](https://stackoverflow.com/questions/54375298/how-to-use-tensorflow-feature-columns-as-input-to-a-keras-model)  and let us know if it helps. Thanks!

**Comment by srfrnk** [comment]
@rmothukuru 
Thanks for your input.
I can't see how the link relates to this issue.
I don't see they use load/save in any way and other then that the code very similar. 
Could you please elaborate?

**Comment by rmothukuru** [comment]
Could reproduce the issue with TF Version 2.0. Here is the [Gist](https://colab.sandbox.google.com/gist/rmothukuru/27bd49c7c30c290e14324bbe1e634787/33085.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
@srfrnk I think this is not related to model saving. I have tried with the simple and official tensorflow model, and I dont see any issue in loading, training, and saving multiple times. Please check the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/eaa7d0c57faa6016016164e9e72f56da/keras_model_saving_multipletimes.ipynb).

Having said that, I think this is more related to usage of `tf.data.dataset` or `FeatureLayer`. Thanks!

**Comment by srfrnk** [comment]
@jvishnuvardhan 
Can you suggest a workaround?
As far as I understand both `tf.data.dataset` and `FeatureLayer` are still supported in TF2.0 am I missing something?
Should saving and loading a model using these not be consistent?

**Comment by raminia** [comment]
I can verify that I have the same issue using a model with DenseFeatures with feature_columns.

**Comment by kevinykuo** [comment]
Can confirm this reproduces in TF2.0 on macOS, and that the second fit call isn't necessary to reproduce, i.e. save->load->save->load breaks.

**Comment by jvishnuvardhan** [comment]
@kevinykuo Did you use `tf.data.dataset` and `Feature_columns` in your code? Can you please share a standalone code to reproduce the issue? Thanks!

**Comment by zmjjmz** [comment]
I'm encountering this issue with TF 2.1 and getting a different traceback from the ones posted here, but this is just on trying to save twice. Check out this [gist](https://gist.github.com/zmjjmz/ade6ecb5eaf011788b1da62db4413cad)

Not sure if being on Python 3.5 is a problem, but I'm also using feature columns here. While saving twice doesn't seem like a real usecase, it comes up inherently in some DAG caching logic this model is a part of.

**Comment by goldiegadde** [comment]
@srfrnk using the colab gist posted [here](https://github.com/tensorflow/tensorflow/issues/33085#issuecomment-541718214)
and tensorflow==2.2.0-rc1, I do not see the multiple save/load issue anymore. 

Can you please verify on your end and close this issue if it works for your use case ?

@zmjjmz please file a new GitHub issue with a standalone repro snippet. Thanks!

**Comment by goldiegadde** [comment]
@srfrnk closing this issue as we believe it is fixed with 2.2.0-rc1 and the latest tf-nightly. PLease feel free to reopen if the issue is not resolved for you. Thanks!

**Comment by ysyyork** [comment]
I'm trying it with today's (apr 1st 2020) nightly build. Still seeing this issue.
